File: cookbook/cpp/data-structures/disjoint-set/disjoint_set.cpp
#include <numeric>
#include <stdexcept>
#include <vector>

class DisjointSet {
public:
    explicit DisjointSet(int element_count)
        : parent_(element_count), component_size_(element_count, 1) {
        if (element_count < 0) {
            throw std::invalid_argument("element_count must not be negative");
        }
        std::iota(parent_.begin(), parent_.end(), 0);
    }

    int find(int element) {
        validate(element);
        if (parent_[element] != element) {
            parent_[element] = find(parent_[element]);
        }
        return parent_[element];
    }

    bool unite(int left, int right) {
        int left_root = find(left);
        int right_root = find(right);
        if (left_root == right_root) {
            return false;
        }

        if (component_size_[left_root] < component_size_[right_root]) {
            std::swap(left_root, right_root);
        }
        parent_[right_root] = left_root;
        component_size_[left_root] += component_size_[right_root];
        return true;
    }

    bool connected(int left, int right) {
        return find(left) == find(right);
    }

    int component_size(int element) {
        return component_size_[find(element)];
    }

private:
    void validate(int element) const {
        if (element < 0 || element >= static_cast<int>(parent_.size())) {
            throw std::out_of_range("element is outside the disjoint set");
        }
    }

    std::vector<int> parent_;
    std::vector<int> component_size_;
};

File: labs/algorithms/python-implementations/algorithms/disjoint_set.py
"""Disjoint-set union with path compression and union by size."""


class DisjointSet:
    """Maintain connected components for integer elements in ``range(size)``."""

    def __init__(self, size: int) -> None:
        if not isinstance(size, int) or isinstance(size, bool):
            raise TypeError("size must be an integer")
        if size < 0:
            raise ValueError("size must be non-negative")
        self._parent = list(range(size))
        self._component_size = [1] * size
        self._components = size

    def __len__(self) -> int:
        return len(self._parent)

    @property
    def components(self) -> int:
        """Return the current number of connected components."""
        return self._components

    def _validate(self, element: int) -> None:
        if not isinstance(element, int) or isinstance(element, bool):
            raise TypeError("element must be an integer")
        if element < 0 or element >= len(self._parent):
            raise IndexError("element is outside this disjoint set")

    def find(self, element: int) -> int:
        """Return the representative and compress the path to it."""
        self._validate(element)
        root = element
        while root != self._parent[root]:
            root = self._parent[root]
        while element != root:
            parent = self._parent[element]
            self._parent[element] = root
            element = parent
        return root

    def union(self, left: int, right: int) -> bool:
        """Merge two components; return whether a merge occurred."""
        left_root = self.find(left)
        right_root = self.find(right)
        if left_root == right_root:
            return False
        if self._component_size[left_root] < self._component_size[right_root]:
            left_root, right_root = right_root, left_root
        self._parent[right_root] = left_root
        self._component_size[left_root] += self._component_size[right_root]
        self._components -= 1
        return True

    def connected(self, left: int, right: int) -> bool:
        return self.find(left) == self.find(right)

    def component_size(self, element: int) -> int:
        return self._component_size[self.find(element)]
